xchange-stream-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingAdapters.java

```diff
@@ -19,58 +15,93 @@
-              for (String key : BID_KEYS) {
-                if (currentNode.has(key)) {
-                  adaptLimitOrders(instrument, Order.OrderType.BID, currentNode.get(key))
-                      .forEach(orderBook::update);
+    public static OrderBook adaptOrderbookMessage(int depth, TreeSet<LimitOrder> bids, TreeSet<LimitOrder> asks, Instrument instrument, ArrayNode arrayNode) {
+        final AtomicLong expectedChecksum = new AtomicLong(0);
+        final AtomicReference<Date> lastTime = new AtomicReference<>(Date.from(Instant.EPOCH));
+        final boolean awaitingSnapshot = (bids.isEmpty() && asks.isEmpty());
+        arrayNode.elements().forEachRemaining(currentNode -> {
+            if (awaitingSnapshot) {
+                if (currentNode.has(BID_SNAPSHOT) && currentNode.has(ASK_SNAPSHOT)) {
+                    LOG.info("Received {} snapshot, clearing book", instrument);
+                    updateInBook(depth, instrument, Order.OrderType.BID, currentNode, BID_SNAPSHOT, bids);
+                    updateInBook(depth, instrument, Order.OrderType.ASK, currentNode, ASK_SNAPSHOT, asks);
                 }
-              }
-              for (String key : ASK_KEYS) {
-                if (currentNode.has(key)) {
-                  adaptLimitOrders(instrument, Order.OrderType.ASK, currentNode.get(key))
-                      .forEach(orderBook::update);
+            } else {
+                if (currentNode.has(BID_UPDATE)) {
+                    updateInBook(depth, instrument, Order.OrderType.BID, currentNode, BID_UPDATE, bids);
                 }
-              }
-            });
-    return new OrderBook(
-        orderBook.getTimeStamp(),
-        Lists.newArrayList(orderBook.getAsks()),
-        Lists.newArrayList(orderBook.getBids()),
-        true);
-  }
+                if (currentNode.has(ASK_UPDATE)) {
+                    updateInBook(depth, instrument, Order.OrderType.ASK, currentNode, ASK_UPDATE, asks);
+                }
+            }
+            if (!awaitingSnapshot && currentNode.has(CHECKSUM)) {
+                expectedChecksum.set(currentNode.get(CHECKSUM).asLong());
+            }
+        });
+        if ( bids.isEmpty() && asks.isEmpty()){
+            LOG.info("Ignoring {} message {}, awaiting snapshot", instrument, arrayNode);
+        }
+        OrderBook result = new OrderBook(lastTime.get(), Lists.newArrayList(asks), Lists.newArrayList(bids), true);
+        long localChecksum = createCrcChecksum(asks, bids);
+        if (expectedChecksum.get() > 0 && expectedChecksum.get() != localChecksum) {
+            LOG.warn("{} checksum does not match, expected {} but local checksum is {}", instrument, expectedChecksum.get(), localChecksum);
+            throw new IllegalStateException("Checksum did not match");
+        } else if (expectedChecksum.get() == 0) {
+            LOG.debug("Skipping {} checksum validation, no expected checksum in message", instrument);
+        }
+        return result;
```

Comment:
you can just
```
return new OrderBook(lastTime.get(), Lists.newArrayList(asks), Lists.newArrayList(bids), true);
```
instead of creating it on line 84